beginner_source/translation_transformer.py

```diff
@@ -0,0 +1,399 @@
+    mask = mask.float().masked_fill(mask == 0, float('-inf')).masked_fill(mask == 1, float(0.0))
+    return mask
+
+def create_mask(src, tgt):
+  src_seq_len = src.shape[0]
+  tgt_seq_len = tgt.shape[0]
+
+  tgt_mask = generate_square_subsequent_mask(tgt_seq_len)
+  src_mask = torch.zeros((src_seq_len, src_seq_len), device=DEVICE).type(torch.bool)
+
+  src_padding_mask = (src == PAD_IDX).transpose(0, 1)
+  tgt_padding_mask = (tgt == PAD_IDX).transpose(0, 1)
+  return src_mask, tgt_mask, src_padding_mask, tgt_padding_mask
+
+
+######################################################################
+# Define model parameters and instantiate model 
+#
+
+SRC_VOCAB_SIZE = len(de_vocab)
+TGT_VOCAB_SIZE = len(en_vocab)
+EMB_SIZE = 512
+NHEAD = 8
+FFN_HID_DIM = 512
+BATCH_SIZE = 128
+NUM_ENCODER_LAYERS = 3
+NUM_DECODER_LAYERS = 3
+NUM_EPOCHS = 20
+
+DEVICE = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
+
+transformer = Seq2SeqTransformer(NUM_ENCODER_LAYERS, NUM_DECODER_LAYERS, 
+                                 EMB_SIZE, SRC_VOCAB_SIZE, TGT_VOCAB_SIZE,
+                                 FFN_HID_DIM)
+
+for p in transformer.parameters():
+    if p.dim() > 1:
+        nn.init.xavier_uniform_(p)
+
+transformer = transformer.to(device)
+
+loss_fn = torch.nn.CrossEntropyLoss(ignore_index=PAD_IDX)
+
+optimizer = torch.optim.Adam(
+    transformer.parameters(), lr=0.0001, betas=(0.9, 0.98), eps=1e-9
+)
+
+######################################################################
+#
+
+def train_epoch(model, train_iter, optimizer):
+  model.train()
+  losses = 0
+  for idx, (src, tgt) in enumerate(train_iter):
+      src = src.to(device)
+      tgt = tgt.to(device)
+            
+      tgt_input = tgt[:-1, :]
+
+      src_mask, tgt_mask, src_padding_mask, tgt_padding_mask = create_mask(src, tgt_input)
+
+      logits = model(src, tgt_input, src_mask, tgt_mask,
+                                src_padding_mask, tgt_padding_mask, src_padding_mask)
+      
+      optimizer.zero_grad()
+      
+      tgt_out = tgt[1:,:]
+      loss = loss_fn(logits.reshape(-1, logits.shape[-1]), tgt_out.reshape(-1))
+      loss.backward()
+
+      optimizer.step()
+      losses += loss.item()
+  return losses / len(train_iter)
+
+
+def evaluate(model, val_iter):
+  model.eval()
+  losses = 0
+  for idx, (src, tgt) in (enumerate(valid_iter)):
+    src = src.to(device)
+    tgt = tgt.to(device)
+
+    tgt_input = tgt[:-1, :]
+
+    src_mask, tgt_mask, src_padding_mask, tgt_padding_mask = create_mask(src, tgt_input)
+
+    logits = model(src, tgt_input, src_mask, tgt_mask,
+                              src_padding_mask, tgt_padding_mask, src_padding_mask)
+    tgt_out = tgt[1:,:]
+    loss = loss_fn(logits.reshape(-1, logits.shape[-1]), tgt_out.reshape(-1))
+    losses += loss.item()
+  return losses / len(val_iter)
+
+######################################################################
+# Train model 
```

Comment:
Can you include the printout of the training process here? Since the transformer based MT outperforms the old RNN based MT in term of performance and accuracy, you can also mention them here as well. Just my 2c.